### PR TITLE
Feature/issue 144 cache profile and convos

### PR DESCRIPTION
**Pull request by edouardmichelin**
Implementing cache system to maintain persistency of the data through network losses.
Not all data will be stored in local cache: only user profile, conversations with their messages and favorite listings.

Also implementing network manager which allows us to get the current status of the network.
On connectivity loss, a message is displayed on the screen of the user to alert them.

#### Code structure of the mainly impacted code

`app.data.local` => Cache config, DAOs (Data Access Object) and everything related to cache
`app.data.network` => Network "manager"
`app.data.repository` => Repository implementations which were updated to make use of the local storage
`app.domain.model` => Domain models which were updated to be storable in the local db

#### Note for reviewer(s)

The big part of the code is in `app.data.{repository, local, network}` and in the tests.
Other modifications (e.g. `app.presentation`) are minor updates related to the updated/added code.

### CODE REVIEW

**Comment by edouardmichelin** [comment]
Coverage is at 60% because I had to modify every repository implementation's functions (in order to add the connection status check and other things) and some of them were untested.
Therefore these untested functions impact the diff-coverage negatively even though the coverage is sufficient on newly added code.

File: app/src/main/java/com/studhub/app/data/local/dao/UserFavListingsDao.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package com.studhub.app.data.local.dao
+
+import androidx.room.*
+import com.studhub.app.data.local.entity.UserFavoriteListings
+import com.studhub.app.domain.model.Listing
+
+@Dao
+interface UserFavListingsDao {
+    @Transaction
+    @Query(
```
Comment (Edouard-6) [inline app/src/main/java/com/studhub/app/data/local/dao/UserFavListingsDao.kt:10]: Nice to see some sql queries in this project.

File: app/src/main/java/com/studhub/app/domain/model/Listing.kt [hunk 2]
```diff
@@ -1,15 +1,31 @@
...
-    val picturesUri: List<Uri>? = null
-)
+    @PrimaryKey
+    @ColumnInfo(name = "listing_id")
+    var id: String = "",
+    var name: String = "",
+    var description: String = "",
+    var sellerId: String = "",
+    var createdAt: Date = Date(),
+    @Ignore
```
Comment (Edouard-6) [inline app/src/main/java/com/studhub/app/domain/model/Listing.kt:19]: Why are you ignoring those?

File: app/src/main/java/com/studhub/app/domain/model/Message.kt [hunk 3]
```diff
@@ -1,14 +1,25 @@
...
-    val id: String = "",
-    val conversationId: String = "",
-    val conversation: Conversation? = null,
-    val senderId: String = "",
-    val sender: User? = null,
-    val content: String = "",
-    val image: String? = null,
-    val createdAt: Date = Date()
-)
+    @PrimaryKey
```
Comment (Edouard-6) [inline app/src/main/java/com/studhub/app/domain/model/Message.kt:11]: Good idea to annotate primary keys this way.

File: app/src/main/java/com/studhub/app/domain/repository/UserRepository.kt [hunk 4]
```diff
@@ -12,12 +12,12 @@ interface UserRepository {
...
     suspend fun addFavoriteListing(
         userId: String,
-        favListingId: String
+        favListing: Listing
     ): Flow<ApiResponse<User>>
 
     suspend fun removeFavoriteListing(
         userId: String,
-        favListingId: String
+        favListing: Listing
```
Comment (Edouard-6) [inline app/src/main/java/com/studhub/app/domain/repository/UserRepository.kt:20]: Which attribute do you need from the listing other than the ID to pass the whole Listing object?

**Review by Edouard-6 — APPROVED** [review]
Perfect refactoring, good cache system for user profile, conversations with their messages and favorite listings. Easy and modular way of implementing network manager. LGTM.
Should improve test coverage to be minimum 80%.

File: app/src/androidTest/java/com/studhub/app/data/local/dao/ConversationDaoTest.kt [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    fun setAndGetMessage() {
+        val conversationId = Random.nextLong().toString()
+        val conversation1 = Conversation(id = conversationId, user1Name = "Bobby")
+        val conversation2 = Conversation(id = Random.nextLong().toString(), user1Name = "Michelle")
+
+        runBlocking {
+            cache.saveConversation(conversation1)
+            cache.saveConversation(conversation2)
+        }
+
```
Comment (tepoorte) [inline app/src/androidTest/java/com/studhub/app/data/local/dao/ConversationDaoTest.kt:50]: nice work here! It's very good to use a cache for the conversation testing by using what we saw in the SwEng course!

File: app/src/main/java/com/studhub/app/data/local/dao/ConversationDao.kt [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+import androidx.room.*
+import com.studhub.app.domain.model.Conversation
+
+@Dao
+interface ConversationDao {
+    @Transaction
+    @Query("SELECT * FROM conversation WHERE user1Id = :userId OR user2Id = :userId ORDER BY updatedAt DESC")
+    suspend fun getConversations(userId: String): List<Conversation>
+
```
Comment (tepoorte) [inline app/src/main/java/com/studhub/app/data/local/dao/ConversationDao.kt:11]: Very very nice! You're implementing what we saw in other courses in order to create the conversation interface!

**Review by tepoorte — APPROVED** [review]
Very important PR here for the entire application, conventions are respected and overall changes and new implementation make the code more readable and understandable. Nice work!

File: app/src/main/java/com/studhub/app/domain/repository/UserRepository.kt (hunk 4 above)
Comment (edouardmichelin) [reply app/src/main/java/com/studhub/app/domain/repository/UserRepository.kt:20]: I need the listing to store it in the cache

**Review by raynasr — COMMENTED** [review]
Huge PR.
Good work.

**Pull request merged**